src/os/bluestore/BlueStore.cc

```diff
@@ -17733,20 +17733,19 @@ int BlueStore::_setattr(TransContext *txc,
 			CollectionRef& c,
 			OnodeRef& o,
 			const string& name,
-			bufferptr& val)
+			bufferlist& val)
 {
   dout(15) << __func__ << " " << c->cid << " " << o->oid
 	   << " " << name << " (" << val.length() << " bytes)"
 	   << dendl;
   int r = 0;
-  if (val.is_partial()) {
-    auto& b = o->onode.attrs[name.c_str()] = bufferptr(val.c_str(),
-						       val.length());
-    b.reassign_to_mempool(mempool::mempool_bluestore_cache_meta);
-  } else {
-    auto& b = o->onode.attrs[name.c_str()] = val;
-    b.reassign_to_mempool(mempool::mempool_bluestore_cache_meta);
+  if (!val.is_contiguous()) {
+    val.rebuild();
+  } else if (val.front().is_partial()) {
+    val.rebuild();
   }
+  auto& b = o->onode.attrs[name.c_str()] = val.front();
```

Comment:
@rzarzynski 
I agree that it is unlikely that attr encoded in `ObjectStore::Transaction` will be perfectly aligned with a `buffer::ptr`.
The exception might be some hand-crafted Transactions created for unit test purposes, but let's ignore those.

Option A:
If we are on simplification run, we could do unconditional
`val.rebuild();`

Option B:
But how about improving
`void decode(buffer::ptr& bp, bufferlist::const_iterator& p)`
so it would directly create buffer::ptr of exact necessary size and copy into it?

Reply:
> Option B:
But how about improving
`void decode(buffer::ptr& bp, bufferlist::const_iterator& p)`
so it would directly create buffer::ptr of exact necessary size and copy into it?

This would affect every instance in the system while nobody expects a deep copy. I guess `unittest_denc` or `unittest_encoding` even check the copy is shallow.